amethyst_animation/src/resources.rs

```diff
@@ -0,0 +1,193 @@
+/// Attach to the root entity that an animation can be defined for.
+/// Only required for animations which targets more than a single node.
+#[derive(Debug, Clone)]
+pub struct AnimationHierarchy {
+    pub nodes: FnvHashMap<usize, Entity>,
+}
+
+impl Component for AnimationHierarchy {
+    type Storage = DenseVecStorage<Self>;
+}
+
+/// Defines a single animation.
+/// Defines relationships between the node index in `AnimationHierarchy` and a `Sampler` handle.
+/// If the animation only targets a single node index, `AnimationHierarchy` is not required.
+#[derive(Clone, Debug)]
+pub struct Animation {
+    /// node index -> sampler handle
+    pub nodes: Vec<(usize, Handle<Sampler>)>,
+}
+
+impl Asset for Animation {
+    type Data = Self;
+    type HandleStorage = DenseVecStorage<Handle<Self>>;
+}
+
+impl Into<Result<Animation>> for Animation {
+    fn into(self) -> Result<Animation> {
+        Ok(self)
+    }
+}
+
+/// State of animation
+#[derive(Debug, Clone, PartialEq)]
+pub enum ControlState {
+    /// Animation was just requested, not started yet
+    Requested,
+    /// Animation is running, contains last animation tick, and accumulated duration
+    Running(Instant, Duration),
+    /// Animation is paused at the accumulated duration
+    Paused(Duration),
+    /// Request termination of the animation
+    Abort,
+    /// Animation is completed
+    Done,
+}
+
+impl ControlState {
+    /// Is the state `Running`
+    pub fn is_running(&self) -> bool {
+        match *self {
+            ControlState::Running(..) => true,
+            _ => false,
+        }
+    }
+
+    /// Is the state `Paused`
+    pub fn is_paused(&self) -> bool {
+        match *self {
+            ControlState::Paused(_) => true,
+            _ => false,
+        }
+    }
+}
+
+/// The rest state for a single attribute
+#[derive(Debug, Clone)]
+pub enum RestState {
+    /// Translation is a 3d vector
+    Translation([f32; 3]),
+    /// Rotation is a quaternion
+    Rotation([f32; 4]),
+    /// Scale is a 3d vector
+    Scale([f32; 3]),
+}
+
+/// Control handling of animation/sampler end
+#[derive(Debug, Clone)]
+pub enum EndControl {
+    /// Loop the requested number of iterations, None = loop infinitely
+    Loop(Option<u32>),
+    /// When duration of sampler/animation is reached, go back to rest state
+    Normal,
+}
+
+/// Run the sampler on the attached entity
+#[derive(Clone)]
+pub struct SamplerControl {
+    /// Sampler
+    pub sampler: Handle<Sampler>,
+    /// State of sampling
+    pub state: ControlState,
+    /// What to do when sampler ends
+    pub end: EndControl,
+    /// What the transform should return to after end
+    pub after: RestState,
+}
+
+/// Sampler control set, containing optional samplers for each of the possible channels.
+///
+/// Note that the `AnimationOutput` in the `Sampler` referenced, and the `RestState` referenced in
+/// each `SamplerControl` should match the attribute channel from the set here. There is no check
```

Comment:
Shouldn't be `SamplerControl` parametrised in this case?

Reply:
This is in part due to how it looked at the start. I did consider changing this note, but came to the conclusion I wanted to get something out first. I will deal with this in a later PR.